email_handler.py

```diff
@@ -536,6 +536,55 @@ def handle_email_sent_to_ourself(alias, from_addr: str, msg: Message, user):
     )
 
 
+def apply_dmarc_policy(alias: Alias, contact: Contact, msg: Message) -> Optional[str]:
+    spam_result = msg.get_all(headers.SPAMD_RESULT)
+    if not spam_result:
+        return False
+    spam_entries = [entry.strip() for entry in spam_result[-1].split("\n")]
+    for iPos in range(len(spam_entries)):
+        sep = spam_entries[iPos].find("(")
+        if sep > -1:
+            spam_entries[iPos] = spam_entries[iPos][:sep]
+    if "DMARC_POLICY_REJECT" in spam_entries:
+        return status.E519
+    if (
+        "DMARC_POLICY_SOFTFAIL" in spam_entries
+        or "DMARC_POLICY_QUARANTINE" in spam_entries
+    ):
+        add_or_replace_header(msg, headers.SL_DIRECTION, "Forward")
```

Comment:
I wonder if we can re-arrange the code so this dmarc checking happens just before sending the email to the mailbox (i.e. the "sl_sendmail" method)?

Reply:
We can for example do the dmarc check in `forward_email_to_mailbox` and raise an exception to abort all the processing and catch the exception in `handle_forward`

Reply:
Re in case you haven't seen this comment :)

Reply:
In that case if there are multiple mailboxes for an alias we'd be doing the same thing once per mailbox. If the mail fails the dmarc check and we do it before the `forward_email_to_mailbox` we only need to to the check once.

Reply:
Yeah if there are multiple mailboxes, the exception will abort everything at the end forward_email_to_mailbox on the first mailbox. I just want to reuse the header manipulation in `forward_email_to_mailbox`.